maintenance/code-backports.sh

```diff
@@ -0,0 +1,73 @@
+#!/bin/sh
+
+# silence debug output from VCS
+beQuiet="--quiet"
+#beQuiet=""
+version=`basename "$1" | sed s/squid-//`
+vnext=$(( $version + 1 ))
+srcbranch=v$vnext
+
+cd $1
+
+# Dev branch does not receive backports
+test -f .BASE && exit 0
+
+# clean the current workspace
+gitCleanWorkspace ()
+{
+  git clean $beQuiet -xdf --exclude="\.BASE"
+  git checkout $beQuiet -- .
+}
+
+if ! test -d .git ; then
+  echo "ERROR: missing git repository"
+  exit 1;
+fi
+if ! test -d ../squid-$vnext ; then
+  echo "ERROR: missing git repository to backport from"
+  exit 1;
+fi
+if test -e ../squid-$vnext/.BASE ; then
+  srcbranch=`cat ../squid-$vnext/.BASE`
+fi
+
+echo "FOUDN srcbranch=$srcbranch"
+
+# Prepare backports branch for updates
+git checkout $beQuiet v$version-next-backports 2>/dev/null &&
+  gitCleanWorkspace &&
+    git rebase origin/v$version 2>/dev/null || ( git rebase --abort 2>/dev/null ; exit 1 )
+
+# handler on successful backport
+doneCherryPick ()
+{
+  sha="$1"
+  prnum="$2"
+  version="$3"
+
+  git push $beQuiet >/dev/null
+  # TODO automatically remove label from PRs on successful commit+push to vN-next-backports
+}
+
+# Find backports to attempt:
+prlist=`gh pr list -L 1 --repo squid-cache/squid --state closed --label backport-to-v$version | wc -l`
+if test "$prlist" -ne 0; then
+  gh pr list --repo squid-cache/squid --state closed --label backport-to-v$version | while read prnum text; do
+    # find a commit in $srcbranch with " (#$prnum)" subject suffix
+    git log --oneline github/$srcbranch --grep=" (#$prnum)\$" | while read sha rest; do
+      git cherry-pick $sha && doneCherryPick $sha $prnum $version || cherry-pick --abort 2>/dev/null
```

Comment:
Please strip `(#prnum)` sequences from cherry-picked commit titles and add original commit SHA references to backported commit message footers instead. See https://github.com/squid-cache/squid/pull/1308#issuecomment-1474018944 for an illustration of the problem this change request is meant to solve.

Reply:
This change would break use of the gh CLI tool. We cannot do this.

Reply:
> > This change would break use of the gh CLI tool.
> 
> What makes you think that? 

The `gh pr list` used to locate which commits are labeled for backport does not have any SHA information. The only unique value available is PR number - which needs to be cross-referenced to the `title` field of the branch log(s) to find the SHA appropriate to the src-branch.

Reply:
> > > This change would break use of the gh CLI tool.

> > What makes you think that?
 
> The `gh pr list` used to locate which commits are labeled for backport does not have any SHA information.

As a starting point, let's assume that the `gh` command stays the same while addressing this change request. This change request is not about that `gh` command. It is about `git` commands that create a new/backport commit. The problem is not in labeled PR search -- that search can stay the same as far as this change request is concerned -- the problem is in the commit message created by this script using `git` commands.

Reply:
> > > > This change would break use of the gh CLI tool.
> 
> > > What makes you think that?
> 
> > The `gh pr list` used to locate which commits are labeled for backport does not have any SHA information.
> 
> As a starting point, let's assume that the `gh` command stays the same while addressing this change request.

Assume that I do fully understand what you are asking for.

The problem with your request is that enacting it on backports from vN to vN-1 impacts the results of the gh command used to backport from vN-1 to vN-2. Making it impossible to achieve the fundamental purpose of this script: backport a change from v7->v6->v5.

Reply:
> The problem with your request is enacting it on backports from vN to vN-1 impacts the results of the gh command used to backport from vN-1 to vN-2. Making it impossible to achieve the fundamental purpose of this script: backport a change from v7->v6->v5.

I do not understand why addressing this change request would make it _impossible_ to backport a change from v7 to v6 to v5. AFAICT, as long as we always label PRs that need an automated backport, the script can (continue to) use the `gh pr list` command to find PRs labeled for backporting. Once a labeled PR is found, the script can (continue to) use that PR number to find the corresponding commit SHA that needs backporting. The script can then cherry-pick that commit and post the corresponding PR to merge the result. All that _without_ copying the original PR number to commits corresponding to other PRs (i.e. satisfying this change request)! Where do you see a problem that is impossible to solve?

Reply:
> > Alex: Once a labeled PR is found, the script can (continue to) use that PR number to find the corresponding commit SHA that needs backporting.

> Amos: After enacting the requested change: No, SHA not found.

Why not? If one labels a GitHub PR for backporting, and that GitHub PR is officially merged, then I see no reason why the script cannot find that PR commit in the official repository. The commit is obviously there, we know its title, and that title is unique because both manual GitHub PR commits and automated Anubis PR commits contain a unique PR ID in their title.

If you are thinking of backporting commits that do not have dedicated PRs, then they do not seem to be applicable to this discussion because those commits cannot be labeled for automatic backporting. If really needed, we can support them as well, but that should be a subject of a very different discussion!

> The `git cherry-pick` preserves the title including origin PR ID.

The result of that command depends on how that command is used. The script can use that command _and_ make sure that the unique PR ID is _not_ copied to the new/resulting commit title. The script fully controls what goes into the backporting PR! We are not being forced to push anything specific into that PR. We decide what (metadata) goes in and what does not.

> You can replicate the problem by simulating a backport of the changes in PR 715 from v5 branch to v4 branch.

PR 715 is not a valid test case because it is not a PR against a v5 branch: That PR _can_ be labeled to be backported to v4, but it would _not_ be a "v5 to v4" backport. For a change to be automatically backported (through a properly labeled PR -- the subject of this script and this change request!) from vX to vY, that change needs a (labeled) vX PR. In your example, X is 5, but PR 715 is not a v5 PR.

However, the corresponding PR 717 (v5 commit f4ade36 with `5.0.4 (#717)` title) could have been automatically backported from v5 into v4 (after its PR was labeled for such backporting). I see no problem with that.

> The PR has already been merged from v6 to v5 branch. It was merged with a title change equivalent to what you are requesting we always do.

Our definitions of "equivalent" differ: PR 715 title is `Release Prep for 5.0.4 and 4.13`. If that change were to be automatically backported into v5 after addressing my change request, the corresponding v5 _commit_ title would have been `Release Prep for 5.0.4 and 4.13 (#X)` (where `X` is the corresponding new/backporting PR number). There are no v5 commits with titles that start with `Release Prep for 5`.

You are probably hinting at v5 PR 717 (commit f4ade36 with a very different `5.0.4 (#717)` title). I agree that PR 717 could have been labeled for automated v4 backporting (as discussed above).

----

> I have already explained that thrice.

AFAICT, the previous responses contain repeated _assertions_ (that X is impossible) rather than explanations (of why X is impossible). That lack of cooperation hinders progress. Since I believe that the "impossible" or "cannot do that" assertion itself is false, I see only two ways to make progress: Find the underlying mistake that led to the wrong conclusion or implement the requested changes myself. The former requires detailing/explaining the assertion, triggering my questions. The latter takes even more time and may require adjustments that may create even more conflicts, so I am avoiding that option.